File: Programmers/DFS_BFS/target_number.py
# Target number solution
# Idea 1
def permute(n, r) -> int:
    k = 1
    j = 1
    for _ in range(r):
        k *= n
        j *= r
        n = n-1
        r = r-1
    return int(k/j)
    

def solve(numbers, target):
    answer = 0
    sum = 0
    r=0
    
    for i in numbers:
        sum += i
    
    while True:
        sum -= numbers[0]
        if sum == target:
            break
        elif sum == 0:
            answer = -1
        r += 1
    
    if answer != -1:
        answer = permute(len(numbers), r)
    
    
    return answer

File: Programmers/DFS_BFS/target_number.cpp
#include <string>
#include <vector>

using namespace std;

int total;
void DFS(vector<int> numbers, int target, int sum, int n) {
	if(n >= numbers.size()){
		if (sum == target) total++;
		return;
	}
	
	DFS(numbers, target, sum + numbers[n], n+1);
	DFS(numbers, target, sum - numbers[n], n+1);
}

int solution(vector<int> numbers, int target) {
	int answer = 0;
	DFS(numbers, target, numbers[0], 1);
	DFS(numbers, target, -numbers[0], 1);
	
	answer = total;
	
	return answer;
}

int main(void){
	vector<int> numbers = {1,1,1,1,1};
	int target = 3;
	
	printf("answer : %d\n", solution(numbers, target));
}
